libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RouteRefresherExecutor.kt

```diff
@@ -0,0 +1,36 @@
+package com.mapbox.navigation.core.routerefresh
+
+import com.mapbox.navigation.base.route.NavigationRoute
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+
+internal interface RouteRefresherProgressCallback {
+
+    fun onStarted()
+
+    fun onResult(routeRefresherResult: RouteRefresherResult)
+}
+
+internal class RouteRefresherExecutor(
+    private val routeRefresher: RouteRefresher,
+    private val scope: CoroutineScope,
+    private val timeout: Long,
+) {
+
+    private var hasCurrentRequest = false
+
+    fun postRoutesToRefresh(
+        routes: List<NavigationRoute>,
+        callback: RouteRefresherProgressCallback
+    ) {
+        scope.launch {
+            if (!hasCurrentRequest) {
+                hasCurrentRequest = true
+                callback.onStarted()
+                val result = routeRefresher.refresh(routes, timeout)
+                callback.onResult(result)
+                hasCurrentRequest = false
+            }
+        }
+    }
+}
```

Comment:
I see a couple of problems with this codebase.

First that it's dangerous to leave callbacks without a response. If a piece of code provides a callback, it expects to be called back at some point to resume operations. Now, if the flag is set the callback is ignored.

Second, is this scenario possible?
- Start planned refresh.
- Wait for the interval to elapse.
- While the planned request is in flight, request immediate refresh.
- The immediate refresh will cancel and pause the planned refresh but the  `hasCurrentRequest` flag will never be reset, so the immediate request will be ignored and we end up with a  "deadlock".

Am I missing something?

Reply:
> First that it's dangerous to leave callbacks without a response. If a piece of code provides a callback, it expects to be called back at some point to resume operations. Now, if the flag is set the callback is ignored.

Agreed. I'll fix that.

> Second, is this scenario possible?

It is. Thanks. I'll fix and add a test.